File: mobile/lib/storage/storage.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

/// Thin wrapper around SharedPreferences so we can swap to a real
/// backend later without touching widgets. Mirrors the web app's
/// `lib/storage.ts` module.
class Storage {
  Storage(this._prefs);
  final SharedPreferences _prefs;

  static late Storage instance;

  static Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    instance = Storage(prefs);
  }

  T get<T>(String key, T fallback) {
    final raw = _prefs.getString(key);
    if (raw == null) return fallback;
    try {
      final decoded = jsonDecode(raw);
      return decoded as T;
    } catch (_) {
      return fallback;
    }
  }

  String? getString(String key) => _prefs.getString(key);

  Future<void> set<T>(String key, T value) =>
      _prefs.setString(key, jsonEncode(value));

  Future<void> setString(String key, String value) =>
      _prefs.setString(key, value);

  Future<void> remove(String key) => _prefs.remove(key);
}

class StorageKeys {
  static const theme = 'itr.theme';
  static const currentUser = 'itr.currentUser';
  static const users = 'itr.users';
  static const candidateProfiles = 'itr.candidateProfiles';
  static const jobs = 'itr.jobs';
  static const subscriptions = 'itr.subscriptions';
  // Employer-side prefs (2026-07): saved candidate searches, shortlist,
  // recently viewed. Each is keyed by employer id inside the value blob so
  // multiple employers sharing a device stay isolated.
  static const employerSavedSearches = 'itr.employer.savedSearches';
  static const employerShortlist = 'itr.employer.shortlist';
  static const employerRecentCandidates = 'itr.employer.recentCandidates';
  static String otp(String email) => 'itr.otp.$email';
}

File: web/src/lib/storage.ts
// Thin wrapper around localStorage / sessionStorage so we can swap to a
// real backend later without touching components.

type Storage = "local" | "session";

const store = (s: Storage) => (s === "local" ? localStorage : sessionStorage);

export const get = <T>(key: string, fallback: T, s: Storage = "local"): T => {
  const raw = store(s).getItem(key);
  if (raw === null) return fallback;
  try {
    return JSON.parse(raw) as T;
  } catch {
    return fallback;
  }
};

export const set = <T>(key: string, value: T, s: Storage = "local"): void => {
  store(s).setItem(key, JSON.stringify(value));
};

export const remove = (key: string, s: Storage = "local"): void => {
  store(s).removeItem(key);
};

export const KEYS = {
  theme: "itr.theme",
  currentUser: "itr.currentUser",
  users: "itr.users",
  candidateProfiles: "itr.candidateProfiles",
  jobs: "itr.jobs",
  subscriptions: "itr.subscriptions",
  // Employer-side prefs (2026-07): saved candidate searches, shortlist,
  // recently viewed. Keyed by employer id inside the value blob so
  // multiple employers sharing a browser stay isolated.
  employerSavedSearches: "itr.employer.savedSearches",
  employerShortlist: "itr.employer.shortlist",
  employerRecentCandidates: "itr.employer.recentCandidates",
  otp: (email: string) => `itr.otp.${email}`,
} as const;
